### DISCUSSION TITLE
Extending models

### DISCUSSION THREAD

**Discussion by llermaly** [hagopj13/node-express-boilerplate#49] [Q&A] [answered]
Hello! , first of all THANK YOU! I learned a lot using this repo and reading the issues. 

I want to use your code as a good practice but I have some issues creating new models, I wanted to start with a simple "products" model, each user can have (tons of) products.

I can easily create the new model Product and attach the user id to it , but then how I reproduce the verification part? (each user can only mess with its own products)

I'm not sure about how to handle it because for users is tied to the "auth" middleware.  Should I put more logic in that function? something like :

```
    if ((!hasRequiredRights && req.params.userId !== user.id) || prod.userId !== user.id) {
      return reject(new ApiError(httpStatus.FORBIDDEN, 'Forbidden'));
    }
```
Doesnt feel good at all

Thanks!

**Comment by hagopj13** [comment]
Hi @llermaly I'm glad you like the boilerplate and find it useful.

I see your point that doing the authorizations globally in this way is very convenient, but I don't really find it scalable. I would put this logic inside the controller (or in the services layer) of the API that is concerned with editing/viewing the products.

If you realize that this is being used very often, you can also create a custom middleware that handles this logic. You can then attach this middleware to any route you want to behave so.

**Comment by llermaly** [comment, marked as answer]
Any chance to have an example? , I ended doing this but not totally happy:

```
const getProduct = catchAsync(async (req, res) => {
  const product = await productService.getProductById(req.params.productId, req.user._id);
  if (!product) {
    throw new ApiError(httpStatus.NOT_FOUND, 'Product not found');
  }
  res.send(product);
});
```

passing down the user id and validating is created_by

**Reply by hagopj13** [reply]
@llermaly this looks good enough. If a user does not have access on the product, then it is already enough for them to know that the product does not exist.

If you really want to separate the 2 errors (forbidden and not found), then create a separate service for each one of them and use them sequentially in the controller. First check if the product is found (return 404 if not), and then check if the current user has access to it (in this case check if craeted_by = userId), and if not, return 403.

**Reply by llermaly** [reply]
Thanks! , I will mark this as answered and I would REALLY like to see something from you that works better than "good enough" to improve my coding quality.

Thanks for your time :)

**Reply by llermaly** [reply]
I Thought about the 403/404 subject but in theory you are fetching -your- products, so the concept of forbidden should not exist but product exist or not exist under your scope.
